Track variable locations

djolertrk opened the pull request:
Track variable locations at HIR and MASM levels.
Add DebugVar decorator and new debug_info section in the final .masp.

Also, add `miden-debugdump` tool to dump .debug_info section, so we can explore the custom section from the package.

vm part: https://github.com/0xMiden/miden-vm/pull/2471

djolertrk commented:
Goes on top of https://github.com/0xMiden/compiler/pull/812.

djolertrk committed:
Add debug variable location tracking to HIR and MASM

This commit adds debug information support for tracking
variable locations through the compilation pipeline:

- Add debug info representation on HIR level (builtin.dbg_value ops,
  DIExpression, DILocalVariable attributes)
- Handle DebugVars during lowering from HIR to MASM
- Add DebugInfoBuilder for constructing debug metadata
- Add RemoveDeadDebugOps pass to clean up debug ops with dead operands
- Add miden-debugdump tool for inspecting .debug_info sections in MASP
  packages (similar to llvm-dwarfdump)
- Add documentation for the debug info format (docs/DebugInfoFormat.md,
  docs/DebugInfoMetadata.md)
- Add lit tests for debug functionality

Run tests:
$ litcheck lit run --verbose tests/lit/debugdump
$ litcheck lit run --path bin ./tests/lit/debug/

djolertrk commented:
It depends on the types and debug decorator defined in https://github.com/0xMiden/miden-vm/pull/2471.

bitwalker commented:
I've done an initial review of the `miden-vm` PR, and I expect we can get that merged ASAP once all the review feedback is addressed.

djolertrk committed:
Align with latest debug_info in miden-vm

djolertrk committed:
fix: compute correct FMP offset for debug variable locations

- Check if WasmLocal value is on Miden operand stack before assuming
  it's in local memory (emit Stack(pos) if on stack)
- Add post-processing in MasmFunctionBuilder::build() to patch
  Local(idx) with correct FMP offset
- Compute num_wasm_locals = params_in_felts + hir_locals for correct
  offset calculation (DWARF WasmLocal uses WASM indexing where params
  come first)
- Add patch_debug_var_locals_in_block() to recursively patch all
  DebugVar decorators in the MASM body

djolertrk committed:
Introduce Debug Info dialect

Confirm that cfg --> scf preserves debug info

djolertrk committed:
fix: handle debug info ops in transforms and MASM codegen

Debug info operations (debuginfo.value, debuginfo.kill, debuginfo.declare)
are purely observational and were causing panics and assembler validation
errors when they survived through optimization passes into codegen.

- Skip operand scheduling for DebugValue (observational, not consuming)
- Gracefully skip debuginfo ops with no HirLowering in the emitter
- Make SinkOperandDefs and ControlFlowSink debug-info-aware so debug
  uses don't block sinking or prevent dead code elimination
- Strip DebugVar-only procedure bodies that would be rejected by the
  Miden assembler (decorators without instructions have no MAST effect)

djolertrk committed:
debuginfo: migrate DI types to DialectAttribute derive for v0.21.1 compatibility

djolertrk committed:
feat: debug variable location tracking with dedup crash fix

Add end-to-end debug variable location tracking through the compiler
pipeline, and patch miden-vm to fix a crash when basic blocks are
deduplicated during assembly.

djolertrk committed:
fix: rebase recovery — entity alignment

Fix data_offset_align and raw_entity_metadata_layout_for_value_layout
to correctly handle types with alignment > 8

djolertrk committed:
fixup: several bugfixes for var loc

- add FrameBase lowering
- DWARF parsing (parse variable/params names)
- and fix duplicate DebugVar emissions

djolertrk committed:
fixup: remove debug info docs

djolertrk committed:
fixup: debug var loc issues in frontend and codegen

djolertrk committed:
fix: update lit tests for current HIR output and debugdump formats

djolertrk committed:
fix: adapt debugdump to upstream Package.version API change

djolertrk commented:
note that it still depends on the local `miden-vm` (https://github.com/0xMiden/miden-vm/pull/2955 is needed).

djolertrk committed:
chore: fix formatting

djolertrk committed:
fix: post-rebase cleanup for origin/next

djolertrk committed:
fix: register DebugValue as no-op in HIR evaluator

djolertrk committed:
fix: skip debug info ops in liveness analysis

djolertrk committed:
fix: skip location schedule emission for undefined variables

djolertrk commented:
NOTE: It uses https://github.com/walnuthq/miden-vm/tree/fix/debug-var-dedup-crash -- but https://github.com/0xMiden/miden-vm/pull/2955 should be merged soon.

djolertrk committed:
fix: clippy for debug_info.rs

djolertrk committed:
fix: preserve DWARF frame-base variables

djolertrk committed:
fix: run debugdump tests via litcheck

djolertrk committed:
fix: use released miden vm debug APIs

djolertrk committed:
fixup: update tests

djolertrk committed:
fixup: improve debugdump help

djolertrk committed:
fix(debug): emit component core type metadata

bitwalker reviewed (commented):
Just a heads up @djolertrk, but I'm rebasing this PR on another branch of mine, and cleaning up a few things locally. I'll probably close this PR and open a new one with the updated branch - all your commits are preserved of course, but wanted to let you know what's up so you don't burn cycles on this in the meantime.
